### PR TITLE
feat(forge): Add test matching support to foundry.toml

### PR DESCRIPTION
**Pull request by lattejed**
## Motivation

https://github.com/foundry-rs/foundry/issues/1346

`test` matching flags should be supported in `foundry.toml` so complex matching setups can be stored and enabled with different config profiles.

## Solution

Add the matching flags from `test` to the `Config` file, swapping kebab for snake case, e.g., `--match-test` is `match_test` in `foundry.toml`.

`Config` entries are merged with `test` flags. If both config entry and flag are present, the flag takes precedence.

Regex doesn't implement PartialEq (required for `Config` testing) so a wrapper type was added, `serde` was setup to deal with this

### CODE REVIEW

**Review by onbjerg — APPROVED** [review]
LGTM but I want @mattsse to take a look as well

**Comment by lattejed** [comment]
A couple of things I don't like here:

1. Wrapping `Regex` to provide `PartialEq`. The problem is `derive` is all or nothing, so using `Regex` means implementing `PartialEq` for the entire `Config` struct by hand. Equality is needed for testing.
2. Manually merging `Filter` and `Config`. I'm not sure if it's possible to make `Figment` work in this case since the members are different types. `Figment` is pretty light on docs.

File: cli/src/cmd/forge/test.rs [hunk 1]
```diff
@@ -71,6 +71,32 @@ pub struct Filter {
     pub path_pattern_inverse: Option<globset::Glob>,
 }
 
+impl Filter {
+    pub fn with_merged_config(&self) -> Self {
+        let config = Config::load();
+        let mut filter = self.clone();
```
Comment (mattsse) [inline cli/src/cmd/forge/test.rs:77]: since this clones regardless, should the function instead accept `self`?

**Review by mattsse — CHANGES_REQUESTED** [review]
fantastic! only smol nits

while the regex wrapper is a bit unfortunate, it's the best solution here.

mind also adding a `forge_test!` for that, see `config.rs` in the cli/tests folder for reference, something like this but for the path options:

```rust
forgetest!(can_set_gas_price, |prj: TestProject, mut cmd: TestCommand| {
    cmd.set_current_dir(prj.root());

    // explicitly set gas_price
    let config = Config { gas_price: 1337, ..Default::default() };
    prj.write_config(config);

    let config = cmd.config();

    assert_eq!(config.gas_price, 1337);

    let config = prj.config_from_output(["--gas-price", "300"]);
    assert_eq!(config.gas_price, 300);
});
```

File: cli/src/cmd/forge/test.rs (hunk 1 above)
Comment (mattsse) [reply cli/src/cmd/forge/test.rs:77]: I see, makes sense.
naming is fine :)

**Comment by lattejed** [comment]
@mattsse good call on the test. There's actually a bug with `Glob` round-tripping a config to disk and back. 

Do you know who did the heavy lifting for `Config` /  `Figment`? I want to validate some assumptions.

**Comment by mattsse** [comment]
> @mattsse good call on the test. There's actually a bug with `Glob` round-tripping a config to disk and back.
> 
> Do you know who did the heavy lifting for `Config` / `Figment`? I want to validate some assumptions.

shoot :)

**Comment by lattejed** [comment]
Ah, I found it. `Glob` doesn't explicitly handle `Option<Glob>` in it's `serde` impl

**Comment by mattsse** [comment]
good to merge @lattejed ?

**Comment by lattejed** [comment]
Should be good to go @mattsse

**Comment by mattsse** [comment]
fantastic, ty!

**Review by mattsse — APPROVED** [review]

**Pull request merged**